# Wine Quality Prediction Using Machine Learning

# OBJECTIVE

The objective of this project is to develop an accurate and reliable machine learning model that can predict the quality of wine based on its physicochemical properties. By analyzing features such as acidity, sugar content, and alcohol levels, this model aims to assist winemakers in quality control and help consumers make informed purchasing decisions.

# DATA SOURCE

The data for the Wine Quality Prediction project comes from a publicly available dataset known as the white Wine Quality dataset . It includes measurements of various chemical properties of the wines, which are used to predict their quality.
The dataset consists of the following key features:-

Fixed acidity,
Volatile acidity,
Citric acid,
Residual sugar,
Chlorides,
Free sulfur dioxide,
Total sulfur dioxide,
Density,
pH,
Sulphates,
Alcohol,
Quality.

# Importing the libraries

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

# Importing the dataset

In [ ]:
dataset = pd.read_csv('winequality-white.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

In [ ]:
print(X)

[[ 7.    0.27  0.36 ...  3.    0.45  8.8 ]
 [ 6.3   0.3   0.34 ...  3.3   0.49  9.5 ]
 [ 8.1   0.28  0.4  ...  3.26  0.44 10.1 ]
 ...
 [ 6.5   0.24  0.19 ...  2.99  0.46  9.4 ]
 [ 5.5   0.29  0.3  ...  3.34  0.38 12.8 ]
 [ 6.    0.21  0.38 ...  3.26  0.32 11.8 ]]


In [ ]:
print(y)

[6 6 6 ... 6 7 6]


In [ ]:
dataset.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    4898.000000       4898.000000  4898.000000     4898.000000   
mean        6.854788          0.278241     0.334192        6.391415   
std         0.843868          0.100795     0.121020        5.072058   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.300000          0.210000     0.270000        1.700000   
50%         6.800000          0.260000     0.320000        5.200000   
75%         7.300000          0.320000     0.390000        9.900000   
max        14.200000          1.100000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  4898.000000          4898.000000           4898.000000  4898.000000   
mean      0.045772            35.308085            138.360657     0.994027   
std       0.021848            17.007137             42.498065     0.002991   
min       0.009000             2.000000              9.000000     0.987110   
25%       0.036000            23.000000            108.000000     0.991723   
50%       0.043000            34.000000            134.000000     0.993740   
75%       0.050000            46.000000            167.000000     0.996100   
max       0.346000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  4898.000000  4898.000000  4898.000000  4898.000000  
mean      3.188267     0.489847    10.514267     5.877909  
std       0.151001     0.114126     1.230621     0.885639  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.090000     0.410000     9.500000     5.000000  
50%       3.180000     0.470000    10.400000     6.000000  
75%       3.280000     0.550000    11.400000     6.000000  
max       3.820000     1.080000    14.200000     9.000000

In [ ]:
dataset.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')

In [ ]:
dataset.shape

(4898, 12)

# Define Target Variable (y) and Feature Variables (X)

In [ ]:
y = dataset['quality']
X = dataset[['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol']]

# Feature Scalling

In [ ]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X = sc.fit_transform(X)

In [ ]:
y.shape

(4898,)

In [ ]:
X.shape

(4898, 11)

In [ ]:
y

0       6
1       6
2       6
3       6
4       6
       ..
4893    6
4894    5
4895    6
4896    7
4897    6
Name: quality, Length: 4898, dtype: int64

In [ ]:
X

array([[ 1.72096961e-01, -8.17699008e-02,  2.13280202e-01, ...,
        -1.24692128e+00, -3.49184257e-01, -1.39315246e+00],
       [-6.57501128e-01,  2.15895632e-01,  4.80011213e-02, ...,
         7.40028640e-01,  1.34184656e-03, -8.24275678e-01],
       [ 1.47575110e+00,  1.74519434e-02,  5.43838363e-01, ...,
         4.75101984e-01, -4.36815783e-01, -3.36667007e-01],
       ...,
       [-4.20473102e-01, -3.79435433e-01, -1.19159198e+00, ...,
        -1.31315295e+00, -2.61552731e-01, -9.05543789e-01],
       [-1.60561323e+00,  1.16673788e-01, -2.82557040e-01, ...,
         1.00495530e+00, -9.62604939e-01,  1.85757201e+00],
       [-1.01304317e+00, -6.77100966e-01,  3.78559282e-01, ...,
         4.75101984e-01, -1.48839409e+00,  1.04489089e+00]])

# Splitting the dataset into the Training set and Test set

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Training the SVR model on the Training set

In [ ]:
from sklearn.svm import SVC
regressor = SVC()
regressor.fit(X_train, y_train)

SVC()

# Predicting the Test set results

In [ ]:
y_pred = regressor.predict(X_test)


In [ ]:
y_pred.shape

(980,)

In [ ]:
y_pred

array([5, 5, 6, 7, 6, 5, 6, 6, 6, 5, 6, 7, 5, 5, 6, 6, 6, 5, 6, 6, 6, 6,
       6, 6, 6, 7, 5, 5, 6, 6, 5, 6, 6, 6, 5, 6, 6, 7, 7, 6, 6, 6, 6, 6,
       5, 6, 5, 7, 6, 6, 7, 6, 5, 6, 6, 6, 6, 6, 5, 6, 5, 5, 6, 6, 6, 6,
       5, 5, 5, 5, 6, 5, 6, 7, 6, 6, 6, 5, 5, 7, 5, 5, 6, 6, 5, 6, 6, 6,
       6, 6, 6, 6, 5, 7, 5, 6, 5, 5, 6, 7, 6, 5, 6, 6, 5, 7, 6, 6, 6, 5,
       5, 6, 6, 5, 6, 7, 5, 5, 7, 6, 5, 5, 6, 5, 5, 7, 5, 6, 6, 6, 6, 6,
       7, 6, 6, 5, 6, 6, 5, 6, 6, 6, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
       5, 6, 5, 6, 6, 6, 6, 6, 6, 6, 5, 7, 6, 6, 5, 6, 6, 6, 6, 6, 6, 6,
       5, 5, 6, 5, 6, 6, 6, 7, 6, 6, 6, 6, 6, 6, 6, 5, 5, 6, 6, 6, 5, 6,
       6, 6, 6, 6, 5, 6, 6, 6, 6, 5, 7, 6, 5, 7, 7, 6, 6, 7, 6, 6, 7, 7,
       6, 6, 6, 6, 5, 5, 6, 5, 6, 5, 6, 6, 5, 5, 5, 6, 6, 6, 6, 6, 5, 6,
       7, 6, 5, 6, 5, 6, 7, 6, 6, 6, 5, 6, 6, 5, 6, 6, 6, 7, 6, 6, 6, 6,
       5, 6, 6, 6, 6, 5, 7, 6, 5, 6, 5, 5, 6, 6, 5, 6, 6, 5, 5, 6, 6, 6,
       6, 5, 5, 6, 6, 6, 6, 6, 5, 5, 6, 5, 5, 6, 5,

# Evaluating the Model Performance

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
confusion_matrix(y_test,y_pred)


array([[  0,   0,   6,   3,   0,   0],
       [  0,   0,  33,  17,   1,   0],
       [  0,   0, 166, 129,   0,   0],
       [  0,   0,  74, 319,  16,   0],
       [  0,   0,   7, 131,  45,   0],
       [  0,   0,   0,  23,  10,   0]])

In [ ]:
accuracy_score(y_test,y_pred)

0.5408163265306123

In [ ]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           3       0.00      0.00      0.00         9
           4       0.00      0.00      0.00        51
           5       0.58      0.56      0.57       295
           6       0.51      0.78      0.62       409
           7       0.62      0.25      0.35       183
           8       0.00      0.00      0.00        33

    accuracy                           0.54       980
   macro avg       0.29      0.26      0.26       980
weighted avg       0.51      0.54      0.50       980



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# EXPLANATION

In this wine quality prediction project, we employed the Support Vector Classifier (SVC) algorithm from the sklearn library, along with the pandas library for data manipulation. After thorough data exploration and preprocessing, including handling missing values and encoding categorical variables, we trained the SVC model to predict wine quality based on features such as acidity, sugar content, and alcohol level.

Despite our efforts, the model achieved an accuracy of 0.54, indicating moderate performance. While this accuracy can provide some insights, it may not be sufficient for practical use. Further model tuning, feature engineering, or trying different algorithms could potentially improve the accuracy and make the model more reliable for real-world applications.





